File: scripts/run_baseline_tuning.py
"""HANDOFF.md 10장 5번: 베이스라인 PID 수동 튜닝 도구.

┌────────────────────────────────────────────────────────────────────┐
│ 이 스크립트는 게인을 "찾아주지 않는다".                              │
│ 작업자(David)가 게인을 직접 바꿔가며 응답과 지표를 본다.             │
│ HANDOFF.md 5장: 베이스라인 PID 게인 튜닝은 대행 금지 항목이다.       │
│ HANDOFF.md 9장: 자동 튜닝 스크립트로 대체하는 것은 금지다.           │
│ 이 파일에 최적화·격자탐색·pidtune 류를 추가하지 말 것.               │
└────────────────────────────────────────────────────────────────────┘

사용법::

    python scripts/run_baseline_tuning.py --kp 3 --ki 0 --kd 200
    python scripts/run_baseline_tuning.py --kp 3 --ki 0 --kd 200 --step 5
    python scripts/run_baseline_tuning.py            # 게인 없이 실행하면 플랜트 정보만

인자 없이 실행하면 플랜트 개루프 특성만 출력하고 종료한다.

튜닝 시 볼 것 (HANDOFF.md 8장 3항):
  - 지표만 보지 말 것. 타각(3번째 subplot)이 지속 진동하면 그 게인은 버린다.
  - 이 플랜트는 이미 적분기를 갖는다(type 1). 계단 목표의 정상상태 오차는
    P 제어만으로도 0 이 된다. Ki 는 상수 외란 제거용이며, 1단계에는 외란이
    없으므로 Ki 를 키우면 오버슈트와 진동만 늘어난다.
    (근거: tests/test_control.py::test_p_control_alone_has_no_steady_state_error)

채택 후 기록할 것:
  docs/03_baseline_pid.md  게인 값, 시험 조건, 지표표, 응답 그림
  docs/journal.md          버린 시도와 그 이유
  docs/decisions.md        D-005 (베이스라인 게인 채택 근거)
"""

from __future__ import annotations

import argparse
import math
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

from usv_heading import (  # noqa: E402
    PIDGains,
    PIDHeading,
    SimOptions,
    heading_metrics,
    metrics_table,
    nomoto_params,
    simulate_heading,
)

D2R = math.pi / 180.0
R2D = 180.0 / math.pi


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--set", dest="set_name", default="mariner")
    ap.add_argument("--kp", type=float, default=None, help="비례 게인 [-]")
    ap.add_argument("--ki", type=float, default=None, help="적분 게인 [1/s]")
    ap.add_argument("--kd", type=float, default=None, help="미분 게인 [s]")
    ap.add_argument("--step", type=float, default=10.0, help="목표 침로 계단 [deg]")
    ap.add_argument("--band", type=float, default=0.05, help="정정 판정 대역 [-]")
    ap.add_argument("--ss-frac", type=float, default=0.30,
                    help="정상상태 구간 (뒤쪽 비율)")
    ap.add_argument("--save", default="figures/baseline_pid.png",
                    help="그림 저장 경로. 빈 문자열이면 저장 안 함")
    args = ap.parse_args()

    p = nomoto_params(args.set_name)

    if None in (args.kp, args.ki, args.kd):
        _print_plant_info(p)
        return 0

    gains = PIDGains(Kp=args.kp, Ki=args.ki, Kd=args.kd)

    # 시험 조건. 최종 검증 조건 설계는 작업자 담당이다 (HANDOFF 5장, D-008).
    opts = SimOptions(t_end=6 * abs(p.T), psi_ref=args.step * D2R)
    out = simulate_heading(p, PIDHeading(gains, p), opts)
    m = heading_metrics(out, band=args.band, ss_frac=args.ss_frac)

    print(f"\n게인: {gains}   |   계단 {args.step:g} deg, 플랜트 {p.name!r}")
    print()
    print(metrics_table([m], [str(gains)]))
    print()

    if args.save:
        from usv_heading.plotting import plot_heading_response

        Path(args.save).parent.mkdir(parents=True, exist_ok=True)
        import matplotlib
        matplotlib.use("Agg")
        plot_heading_response(
            out,
            ["baseline PID"],
            f"{p.name}: step {args.step:g} deg, {gains}",
            savepath=args.save,
        )
        print(f"그림 저장: {args.save}")

    print()
    print("타각 이력(3번째 subplot)에 지속 진동이 보이면 지표가 좋아도 채택하지 않는다.")
    print("채택 시 docs/03_baseline_pid.md, journal.md, decisions.md D-005 에 기록할 것.")
    print()
    return 0


def _print_plant_info(p) -> None:
    print()
    print("  게인이 지정되지 않았다. --kp / --ki / --kd 를 주고 다시 실행할 것.")
    print()
    print("  이 플랜트의 개루프 특성:")
    print(f"    K = {p.K:+.5f} 1/s,  T = {p.T:+.2f} s")
    print("    전달함수  psi(s)/delta(s) = K / (s*(1 + T*s))")
    if p.actuator.enable:
        print(f"    타각 한계 {p.actuator.delta_max * R2D:.1f} deg, "
              f"변화율 한계 {p.actuator.delta_rate_max * R2D:.1f} deg/s")
    else:
        print("    구동기 제한 없음")
    print()
    print("  분모에 s 가 있으므로 type 1 이다 — 계단 목표의 정상상태 오차는")
    print("  P 제어만으로 0 이 된다. Ki 는 상수 외란(조류·바람) 제거용이다.")
    print()
    print("  어떤 값부터 시작할지는 작업자가 정한다 (HANDOFF.md 5장 — 대행 금지).")
    print()


if __name__ == "__main__":
    raise SystemExit(main())

File: matlab/src/scripts/run_baseline_tuning.m
% RUN_BASELINE_TUNING
%
% HANDOFF.md 10장 5번: 베이스라인 PID 수동 튜닝 도구.
%
% ┌────────────────────────────────────────────────────────────────────┐
% │ 이 스크립트는 게인을 "찾아주지 않는다".                              │
% │ 작업자(David)가 아래 GAINS 값을 직접 바꿔가며 응답과 지표를 본다.    │
% │ HANDOFF.md 5장: 베이스라인 PID 게인 튜닝은 대행 금지 항목이다.       │
% │ HANDOFF.md 9장: 자동 튜닝 스크립트로 대체하는 것은 금지다.           │
% │ 이 파일에 pidtune / fminsearch / 격자탐색 등을 추가하지 말 것.       │
% └────────────────────────────────────────────────────────────────────┘
%
% 사용법:
%   1) 아래 GAINS.Kp, Ki, Kd 에 값을 넣는다
%   2) >> run_baseline_tuning
%   3) 플롯(침로/선수각속도/타각)과 지표표를 본다
%   4) 만족스러울 때까지 1~3 반복
%   5) 채택한 게인과 그렇게 정한 이유를 docs/03_baseline_pid.md 에 기록
%   6) 튜닝 과정에서 버린 시도도 docs/journal.md 에 남긴다
%
% 튜닝 시 볼 것 (HANDOFF.md 8장 3항):
%   - 지표만 보지 말 것. 타각(3번째 subplot)이 지속 진동하면 그 게인은 버린다.
%   - 정상상태 오차가 남으면 Ki, 오버슈트가 크면 Kd 를 먼저 본다.

clear; clc;

%% ---- 1. 튜닝 대상 -----------------------------------------------------
SET_NAME  = 'mariner';
p = nomotoParams(SET_NAME);

d2r = pi/180; r2d = 180/pi;

%% ---- 2. 게인 (작업자가 직접 입력) -------------------------------------
% 단위: deltaCmd[rad] = Kp*e[rad] + Ki*int e[rad*s] + Kd*(-r)[rad/s]
%
% NaN 이면 스크립트가 멈춘다. 값을 넣기 전에는 실행되지 않는다.
GAINS.Kp = NaN;   % [-]
GAINS.Ki = NaN;   % [1/s]
GAINS.Kd = NaN;   % [s]

if any(isnan([GAINS.Kp, GAINS.Ki, GAINS.Kd]))
    fprintf('\n');
    fprintf('  게인이 아직 비어 있다.\n');
    fprintf('  %s 의 GAINS 를 채운 뒤 다시 실행할 것.\n', mfilename('fullpath'));
    fprintf('\n');
    fprintf('  참고 — 이 플랜트의 개루프 특성:\n');
    fprintf('    K = %+.5f 1/s,  T = %+.2f s\n', p.K, p.T);
    fprintf('    전달함수  psi(s)/delta(s) = K / (s*(1 + T*s))\n');
    fprintf('    타각 한계 %.1f deg, 변화율 한계 %.1f deg/s\n', ...
        p.actuator.deltaMax*r2d, p.actuator.deltaRateMax*r2d);
    fprintf('\n');
    fprintf('  게인 값을 무엇으로 시작할지는 작업자가 정한다.\n');
    fprintf('  (HANDOFF.md 5장 — 대행 금지 항목)\n\n');
    return
end

%% ---- 3. 시험 조건 -----------------------------------------------------
% 아래 조건은 "튜닝용" 기본 조건이다.
% 최종 검증 조건 설계는 작업자 담당이다 (HANDOFF.md 5장).
% 확정하면 docs/05_validation.md 에 기록한다.
STEP_DEG = 10;                 % 목표 침로 계단 [deg]
                               % 주의: 선형 Nomoto 의 유효 범위는 작은 타각이다.
                               % docs/02_plant_model.md 의 선형/비선형 비교 참조.

opts = struct( ...
    'tEnd',   6 * abs(p.T), ...
    'h',      abs(p.T) / 500, ...
    'psi0',   0, ...
    'r0',     0, ...
    'delta0', 0, ...
    'psiRef', STEP_DEG * d2r);

%% ---- 4. 시뮬레이션 ----------------------------------------------------
ctrl = @(t, psi, r, psiRef, st, h) pidHeading(t, psi, r, psiRef, st, h, GAINS, p);
out  = simulateHeading(p, ctrl, opts);

%% ---- 5. 지표 및 플롯 --------------------------------------------------
m = headingMetrics(out, struct('band', 0.05, 'ssFrac', 0.30));

fprintf('\n게인: Kp = %.4f, Ki = %.5f, Kd = %.4f   |   계단 %g deg, 플랜트 "%s"\n', ...
    GAINS.Kp, GAINS.Ki, GAINS.Kd, STEP_DEG, p.name);
printMetrics({m}, {sprintf('Kp=%.3g Ki=%.3g Kd=%.3g', GAINS.Kp, GAINS.Ki, GAINS.Kd)});

plotHeadingResponse(out, {'baseline PID'}, ...
    sprintf('%s: 계단 %g deg, Kp=%.3g Ki=%.3g Kd=%.3g', ...
    p.name, STEP_DEG, GAINS.Kp, GAINS.Ki, GAINS.Kd));

%% ---- 6. 다음 단계 안내 ------------------------------------------------
fprintf('타각 이력(3번째 subplot)에 지속 진동이 보이면 지표가 좋아도 채택하지 않는다.\n');
fprintf('채택 시 기록할 것:\n');
fprintf('  docs/03_baseline_pid.md  게인 값, 시험 조건, 위 지표표, 응답 그림\n');
fprintf('  docs/journal.md          버린 시도와 그 이유\n');
fprintf('  docs/decisions.md        D-005 (베이스라인 게인 채택 근거)\n\n');
